engine/src/main/java/io/camunda/zeebe/engine/processing/common/CommandDistributionBehavior.java

```diff
@@ -0,0 +1,79 @@
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.processing.common;
+
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.SideEffectWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.Writers;
+import io.camunda.zeebe.protocol.Protocol;
+import io.camunda.zeebe.protocol.impl.record.UnifiedRecordValue;
+import io.camunda.zeebe.protocol.impl.record.value.distribution.CommandDistributionRecord;
+import io.camunda.zeebe.protocol.record.intent.CommandDistributionIntent;
+import io.camunda.zeebe.stream.api.InterPartitionCommandSender;
+import io.camunda.zeebe.stream.api.records.TypedRecord;
+import io.camunda.zeebe.stream.api.state.KeyGenerator;
+import java.util.List;
+import java.util.stream.IntStream;
+
+public final class CommandDistributionBehavior {
+
+  private final StateWriter stateWriter;
+  private final SideEffectWriter sideEffectWriter;
+  private final List<Integer> otherPartitions;
+  private final InterPartitionCommandSender interPartitionCommandSender;
+  private final KeyGenerator keyGenerator;
+
+  public CommandDistributionBehavior(
+      final Writers writers,
+      final int currentPartition,
+      final int partitionsCount,
+      final InterPartitionCommandSender partitionCommandSender,
+      final KeyGenerator keyGenerator) {
+    stateWriter = writers.state();
+    sideEffectWriter = writers.sideEffect();
+    interPartitionCommandSender = partitionCommandSender;
+    this.keyGenerator = keyGenerator;
+    otherPartitions =
+        IntStream.range(Protocol.START_PARTITION_ID, Protocol.START_PARTITION_ID + partitionsCount)
+            .filter(partition -> partition != currentPartition)
+            .boxed()
+            .toList();
+  }
+
+  public <T extends UnifiedRecordValue> void distributeCommand(final TypedRecord<T> command) {
+    final var distributionRecord =
+        new CommandDistributionRecord()
+            .setValueType(command.getValueType())
+            .setRecordValue(command.getValue());
+
+    stateWriter.appendFollowUpEvent(
+        command.getKey(), CommandDistributionIntent.STARTED, distributionRecord);
+
+    final long key = keyGenerator.nextKey();
+    otherPartitions.forEach(
+        (partition) -> distributeToPartition(command, partition, distributionRecord, key));
+  }
+
+  private <T extends UnifiedRecordValue> void distributeToPartition(
+      final TypedRecord<T> command,
+      final int partition,
+      final CommandDistributionRecord distributionRecord,
+      final long key) {
+    stateWriter.appendFollowUpEvent(
+        command.getKey(),
+        CommandDistributionIntent.DISTRIBUTING,
+        distributionRecord.copy().setPartitionId(partition));
```

Comment:
Some of the records that we'll be distributing may contain a lot of data, e.g. deployments of large resources. To reduce the amount of data we write, we need to slim down the contents of some of these records. We only need the actual `commandValue` for the `STARTED` event (that's when it's stored in the state). So we can remove it from the other places we write the record.

❌ Write the DISTRIBUTING event for the key with a slimmed-down record value: 
```json
{
  "partitionId": 2,
  "valueType": "DEPLOYMENT"
  "commandValue": null
}
```

Reply:
Good catch 👍